File: python/zpe_bio/codec.py
"""
ZPE-Bio Codec — Python Reference Implementation
=================================================
Ratified Phase 1.1 Dual-Mode Reference Implementation.
Supports TRANSPORT (Direction-only) and CLINICAL (Direction + Magnitude) modes.
PRD Reference: ZPE_BIOSIGNAL_CODEC_PRD_v1.0.md §6.5 §4.2
"""

from dataclasses import dataclass
from enum import Enum
from typing import List, Optional, Tuple, Union
import numpy as np

# ── Constants ───────────────────────────────────────────────────────────────
ECG_SAMPLE_RATE = 250
PPG_SAMPLE_RATE = 100
MAX_RLE_COUNT = 65535

# Adaptive Threshold Constants (RUNBOOK_00 §6)
ADAPTIVE_K = 0.15
ADAPTIVE_THR_MIN = 0.001
ADAPTIVE_ALPHA = 0.95

# Weber-Fechner Log-Magnitude Table (64 entries, 6-bit)
# base = 255^(1/63) ≈ 1.091928
LOG_MAG_TABLE = np.array([round(1.091928**i) for i in range(64)], dtype=np.int32)
LOG_BASE = np.log(1.091928)

class CodecMode(Enum):
    TRANSPORT = "transport"  # High compression (CR >= 5x), direction-only
    CLINICAL = "clinical"  # High fidelity (PRD < 5%), direction + magnitude


# Mode-specific presets (normalised mV units)
MODE_CONFIGS = {
    CodecMode.TRANSPORT: {"threshold": 0.050, "step": 0.050},
    CodecMode.CLINICAL: {"threshold": 0.001, "step": 0.001},
}

DIRECTION_DELTAS = {
    0: 0.0,
    1: 1.0,    # +1x
    2: 8.0,    # +8x
    3: 32.0,   # +32x
    4: 128.0,  # +128x
    5: -32.0,  # -32x
    6: -8.0,   # -8x
    7: -1.0,   # -1x
}

@dataclass
class EncodedStream:
    # tokens are (direction, magnitude_idx, count)
    # in TRANSPORT, magnitude_idx is always 0 (val 1)
    # in CLINICAL, magnitude_idx is 0-63 (index into LOG_MAG_TABLE)
    tokens: List[Tuple[int, int, int]]
    start_value: float
    threshold: float
    step: float
    sample_rate: int
    signal_type: str
    thr_mode: str
    mode: CodecMode
    num_samples: int

    @property
    def num_tokens(self) -> int:
        return len(self.tokens)

    @property
    def compression_ratio(self) -> float:
        raw_bits = self.num_samples * 16  # Assuming 16-bit raw samples
        
        # Phase 2.3 Breakthrough VLB:
        # Each Token: 3b direction
        # + mag: 0b (Transport) or 6b (Clinical)
        # + RLE: 1b flag + (4b if short else 16b)
        total_bits = 0
        for d, m_idx, count in self.tokens:
            token_bits = 3 # direction
            if self.mode == CodecMode.CLINICAL:
                token_bits += 6 # mag_idx
            
            # RLE bit packing: 1 bit flag + 4 bits (short) or 16 bits (long)
            token_bits += 1
            if count < 16:
                token_bits += 4
            else:
                token_bits += 16
            total_bits += token_bits
            
        return raw_bits / total_bits if total_bits > 0 else 0.0


def quantise_ecg(delta: float, threshold: float) -> int:
    """Canonical Hyper-Gear selection."""
    abs_delta = abs(delta)
    if abs_delta < threshold:
        return 0

    if delta > 0:
        if abs_delta > 64 * threshold:
            return 4
        if abs_delta > 16 * threshold:
            return 3
        if abs_delta > 4 * threshold:
            return 2
        return 1

    if abs_delta > 16 * threshold:
        return 5
    if abs_delta > 4 * threshold:
        return 6
    return 7


def encode(
    samples: np.ndarray,
    mode: Union[CodecMode, str] = CodecMode.CLINICAL,
    threshold: Optional[float] = None,
    thr_mode: str = "fixed",
    signal_type: str = "ecg",
    step: Optional[float] = None,
) -> EncodedStream:
    if len(samples) < 2:
        raise ValueError("Signal too short")
    
    # Robust String-to-Enum conversion
    if isinstance(mode, str):
        mode = CodecMode(mode.lower())

    fs = ECG_SAMPLE_RATE if signal_type == "ecg" else PPG_SAMPLE_RATE
    mode_cfg = MODE_CONFIGS[mode]
    threshold = threshold if threshold is not None else mode_cfg["threshold"]
    step = step if step is not None else mode_cfg["step"]
    
    samples_f = samples.astype(np.float64)
    raw_tokens = []
    reconstructed_val = samples_f[0]
    
    # Adaptive envelope state
    env = 0.0
    
    for actual in samples_f[1:]:
        delta = actual - reconstructed_val
        
        # Update adaptive threshold if enabled
        if thr_mode == "adaptive_rms":
            # Simple envelope tracker: env = alpha * env + (1-alpha) * |delta|
            env = ADAPTIVE_ALPHA * env + (1.0 - ADAPTIVE_ALPHA) * abs(delta)
            current_thr = max(ADAPTIVE_THR_MIN, ADAPTIVE_K * env)
        else:
            current_thr = threshold
            
        d = quantise_ecg(delta, current_thr)
        
        if mode == CodecMode.TRANSPORT:
            mag_val = 1
            mag_idx = 0
        else:
            if d == 0:
                mag_idx = 0
                mag_val = 0
            else:
                d_delta = DIRECTION_DELTAS[d]
                target_m = abs(delta) / (abs(d_delta) * step)
                
                # Optimized find closest index via log math
                # Standardized Optimized Lookup (matches Rust Linear Search Parity)
                # np.argmin picks the FIRST occurrence of the minimum error.
                mag_idx = int(np.argmin(np.abs(LOG_MAG_TABLE - target_m)))
                mag_val = int(LOG_MAG_TABLE[mag_idx])
        
        raw_tokens.append((d, mag_idx))
        reconstructed_val += (DIRECTION_DELTAS[d] * mag_val) * step
    
    return EncodedStream(
        tokens=compress_rle_hybrid(raw_tokens),
        start_value=float(samples_f[0]),
        threshold=threshold,
        step=step,
        sample_rate=fs,
        signal_type=signal_type,
        thr_mode=thr_mode,
        mode=mode,
        num_samples=len(samples),
    )


def decode(encoded: EncodedStream) -> np.ndarray:
    raw_tokens = decompress_rle_hybrid(encoded.tokens)
    reconstructed = np.zeros(len(raw_tokens) + 1, dtype=np.float64)
    reconstructed[0] = encoded.start_value
    for i, (d, mag_idx) in enumerate(raw_tokens):
        if encoded.mode == CodecMode.TRANSPORT:
            mag_val = 1
        else:
            mag_val = LOG_MAG_TABLE[mag_idx] if d != 0 else 0
        reconstructed[i + 1] = reconstructed[i] + (DIRECTION_DELTAS[d] * mag_val * encoded.step)
    return reconstructed


def compress_rle_hybrid(raw_tokens: List[Tuple[int, int]]) -> List[Tuple[int, int, int]]:
    if not raw_tokens:
        return []
    result = []
    curr_d, curr_m = raw_tokens[0]
    curr_count = 1
    for d, m in raw_tokens[1:]:
        if d == curr_d and m == curr_m and curr_count < MAX_RLE_COUNT:
            curr_count += 1
        else:
            result.append((curr_d, curr_m, curr_count))
            curr_d, curr_m = d, m
            curr_count = 1
    result.append((curr_d, curr_m, curr_count))
    return result


def decompress_rle_hybrid(tokens: List[Tuple[int, int, int]]) -> List[Tuple[int, int]]:
    result = []
    for d, m, count in tokens:
        result.extend([(d, m)] * count)
    return result


def compute_prd(original: np.ndarray, reconstructed: np.ndarray) -> float:
    """PRD calculation with stability floor."""
    x = original.astype(np.float64)
    x_hat = reconstructed.astype(np.float64)
    length = min(len(x), len(x_hat))
    
    # Accuracy floor for very small signals (e.g. Asystole)
    if np.max(np.abs(x[:length])) < 1e-4:
        return 0.0 if np.max(np.abs(x_hat[:length])) < 1e-4 else 100.0
        
    numerator = np.sqrt(np.sum((x[:length] - x_hat[:length]) ** 2))
    denominator = np.sqrt(np.sum((x[:length] - np.mean(x[:length])) ** 2))
    return 100.0 * numerator / denominator if denominator > 1e-8 else 0.0


def compute_rmse(original: np.ndarray, reconstructed: np.ndarray) -> float:
    """Standardized RMSE / range."""
    x = original.astype(np.float64)
    x_hat = reconstructed.astype(np.float64)
    length = min(len(x), len(x_hat))
    err = np.sqrt(np.mean((x[:length] - x_hat[:length]) ** 2))
    s_range = x[:length].max() - x[:length].min()
    return err / s_range if s_range > 1e-8 else 0.0

File: core/rust/src/codec.rs
use crate::quantise::{quantise_ecg, DIRECTION_DELTAS};
use heapless::Vec;

#[derive(Clone, Copy, PartialEq, Debug)]
pub enum CodecMode {
    Transport,
    Clinical,
}

#[derive(Clone, Copy, PartialEq, Debug)]
pub enum ThresholdMode {
    Fixed,
    AdaptiveRms,
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum CodecError {
    InputTooShort,
    EncodeBufferFull,
    EmptyOutputBuffer,
}

pub const ADAPTIVE_K: f64 = 0.15;
pub const ADAPTIVE_THR_MIN: f64 = 0.001;
pub const ADAPTIVE_ALPHA: f64 = 0.95;
pub const LOG_BASE: f64 = 0.08794406087123912; // ln(1.091928)

pub const LOG_MAG_TABLE: [f64; 64] = [
    1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 5.0,
    5.0, 6.0, 6.0, 7.0, 8.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 17.0, 18.0, 20.0, 22.0,
    24.0, 26.0, 28.0, 31.0, 34.0, 37.0, 40.0, 44.0, 48.0, 52.0, 57.0, 62.0, 68.0, 74.0, 81.0, 89.0,
    97.0, 106.0, 115.0, 126.0, 138.0, 150.0, 164.0, 179.0, 196.0, 214.0, 233.0, 255.0,
];

pub struct EncodedStream<const N: usize> {
    pub tokens: Vec<(u8, u8, u16), N>,
    pub start_value: f64,
    pub threshold: f64,
    pub step: f64,
    pub mode: CodecMode,
    pub thr_mode: ThresholdMode,
    pub num_samples: usize,
}

impl<const N: usize> EncodedStream<N> {
    pub fn compression_ratio(&self) -> f64 {
        let raw_bits = (self.num_samples * 16) as f64;
        let mut total_bits = 0;
        for &(_d, _m, count) in &self.tokens {
            let mut token_bits = 3; // direction
            if self.mode == CodecMode::Clinical {
                token_bits += 6; // log_magnitude
            }
            token_bits += 1; // RLE flag
            if count < 16 {
                token_bits += 4;
            } else {
                token_bits += 16;
            }
            total_bits += token_bits;
        }
        if total_bits > 0 {
            raw_bits / (total_bits as f64)
        } else {
            0.0
        }
    }
}

pub fn encode<const N: usize>(
    samples: &[f64],
    mode: CodecMode,
    threshold: f64,
    thr_mode: ThresholdMode,
    step: f64,
) -> Result<EncodedStream<N>, CodecError> {
    if samples.len() < 2 {
        return Err(CodecError::InputTooShort);
    }

    let mut tokens = Vec::new();
    let mut reconstructed_val = samples[0];

    let mut current_d = 0u8;
    let mut current_m = 0u8;
    let mut current_count = 0u16;

    let mut env = 0.0;

    for &actual in &samples[1..] {
        let delta = actual - reconstructed_val;

        let current_thr = if thr_mode == ThresholdMode::AdaptiveRms {
            env = ADAPTIVE_ALPHA * env + (1.0 - ADAPTIVE_ALPHA) * delta.abs();
            let adapt = ADAPTIVE_K * env;
            if adapt > ADAPTIVE_THR_MIN {
                adapt
            } else {
                ADAPTIVE_THR_MIN
            }
        } else {
            threshold
        };

        let d = quantise_ecg(delta, current_thr);

        let (mag_idx, mag_val) = if mode == CodecMode::Transport {
            (0u8, 1.0)
        } else if d == 0 {
            (0u8, 0.0)
        } else {
            let d_delta = DIRECTION_DELTAS[d as usize];
            let target_m = delta.abs() / (d_delta.abs() * step);

            // Find closest index in LOG_MAG_TABLE (Standardized Linear Search)
            let mut best_idx = 0;
            let mut min_err = (LOG_MAG_TABLE[0] - target_m).abs();
            for (j, value) in LOG_MAG_TABLE.iter().enumerate().skip(1) {
                let err = (*value - target_m).abs();
                if err < min_err {
                    min_err = err;
                    best_idx = j;
                }
            }
            (best_idx as u8, LOG_MAG_TABLE[best_idx])
        };

        if current_count > 0 && d == current_d && mag_idx == current_m && current_count < u16::MAX {
            current_count += 1;
        } else {
            if current_count > 0 {
                tokens
                    .push((current_d, current_m, current_count))
                    .map_err(|_| CodecError::EncodeBufferFull)?;
            }
            current_d = d;
            current_m = mag_idx;
            current_count = 1;
        }

        reconstructed_val += (DIRECTION_DELTAS[d as usize] * mag_val) * step;
    }

    if current_count > 0 {
        tokens
            .push((current_d, current_m, current_count))
            .map_err(|_| CodecError::EncodeBufferFull)?;
    }

    Ok(EncodedStream {
        tokens,
        start_value: samples[0],
        threshold,
        step,
        mode,
        thr_mode,
        num_samples: samples.len(),
    })
}

pub fn decode_into<const N: usize>(
    encoded: &EncodedStream<N>,
    output: &mut [f64],
) -> Result<usize, CodecError> {
    if output.is_empty() {
        return Err(CodecError::EmptyOutputBuffer);
    }
    output[0] = encoded.start_value;
    let mut idx = 1;
    let mut val = encoded.start_value;
    for &(d, m_idx, count) in &encoded.tokens {
        let mag_val = if encoded.mode == CodecMode::Transport {
            1.0
        } else if d == 0 {
            0.0
        } else {
            LOG_MAG_TABLE[m_idx as usize]
        };

        let delta = (DIRECTION_DELTAS[d as usize] * mag_val) * encoded.step;
        for _ in 0..count {
            if idx >= output.len() {
                break;
            }
            val += delta;
            output[idx] = val;
            idx += 1;
        }
    }
    Ok(idx)
}
